### DISCUSSION TITLE
Implemented Command Palette in pre-release

### DISCUSSION THREAD

**Discussion by kjk** [sumatrapdfreader/sumatrapdf#2476] [Announcements]
I just added command palette in [pre-release builds](https://www.sumatrapdfreader.org/prerelease).

I'm pretty excited about this feature. Use `Ctrl + K` to invoke the palette:

<p align="center">
  <img  src="https://user-images.githubusercontent.com/7935/162860026-cc575787-1f6e-4a16-8a22-bc130e1791dd.png">
</p>

Command palette has recently become popular. See e.g. Visual Studio Code.

How to use it:
* press `Ctrl-K` to show command palette window
* enter text to narrow down list of matches
* up / down arrow will navigate between matches
* `Enter` to execute selected match or double-click in list
* `Escape` to close the window (or click outside of it)

The list contains:
* list of currently opened files
* list of files from history
* commands

You can use it:
* quickly open any file you previously opened
* quickly switch to another tab
* quickly execute any command

It's not perfect yet but it works.

**Comment by pritkr** [comment]
TabWidth setting from advanced options does not work in the pre-release build
I don't know if this is related to this

**Reply by kjk** [reply]
It wasn't. 
I recently added sanity checks for some settings and made `TabWidth` a minimum of 300. 
That is excessive so I lowered it to 60  in https://github.com/sumatrapdfreader/sumatrapdf/commit/647b860ac52f7a099aa848aa0e8360538b87320f

**Comment by 2tef** [comment]
Noticed the last option says "Toggle fullscree**m**" instead of "Toggle fullscreen", and maybe "Check for Update" should be "Check for Update**s**"? hope that helps, love the new command palette!

**Reply by GitHubRulesOK** [reply]
well spotted so made me look closer 

@kjk `Create  Free Text Annotation` is naturally out of alpha order due to double space before free :-)
 also `Deubg:` :-)

**Reply by kjk** [reply]
thanks, fixed in ed2f714eb810387702eaa408a63787b8d24714f7

**Comment by drdaxxy** [comment]
Thanks, this could help a lot with the "too many tabs" problem.

When paths get too long, though, they get cut off:
![screenshot of Command Palette with overflowing paths](https://user-images.githubusercontent.com/40321/167647681-aff089c5-51a7-45d8-adc1-5a1bd8a34277.png)

The instant search UIs in editors usually solve this by printing file names as the main item, with the full path in a smaller font to the right or below. Just putting it in parentheses after the file name would be simpler and solve the problem, IMO.

On that note,
```
   // avoid adding the same file opened in multiple window
```
if I deliberately open a file multiple times, it's because I want to quickly jump between scroll positions, if I do it accidentally, I'd like to see my mistake in the open document list, so this behaviour isn't ideal, but that is a nitpick 😄
